[DNM][JOSS] JOSS paper (#95)
jjhelmus/pyfive: paper.md

# Statement of need

HDF5 is probably the most important data format in physical science, used across the piste.It is particularly important in environmental science, particularly given the fact that NetCDF4 is HDF5 under the hood. From satellite missions, to climate models and radar systems, the default binary format has been HDF5 for decades. While newer formats are starting to get mindshare, there are petabytes, if not exabytes of existing HDF5, and there are still many good use-cases for creating new data in HDF5. However, despite the history, there are few libraries for reading HDF5 file data that do not depend on the official HDF5 library maintained by the HDFGroup, and in particular, there are none that can be used with Python. 
While  the HDF5 c-library is reliable and performant, and battle-tested over decades, there are some caveats to depending upon it: Firstly, it is not thread-safe, and secondly, the code is large and complex, and should anything happen to the financial stability of The HDF5group, it is not obvious the C-code could be maintained. From a long-term curation perspective this last constraint is a concern.

> The original implementation of pyfive (by JH and BM), which included all the low-level functionality to deal with the internals of an HDF5 file was developed with POSIX access in mind. The recent upgrades were developed with the use-case of performant remote access to curated data as the primary motivation, but with additional motivations of having a lightweight HDF5 reader capable of deploying in resource or operating-system constrained environments (such as mobile), and one that could be maintained long-term as a reference reader for curation purposes. The lightweight deployment consequences of a pure-python HDF5 reader need no further introduction, but as additional motivation we now expand on the issues around remote access and curation.

Author: many thanks @bmaranville :beer:

   index: 2
   ror: 05v62cm79
 - name: Instituto de Física de Cantabria (IFCA), CSIC-Universidad de Cantabria, Santander, Spain.
   index: 3
 - name:  European Synchrotron Radiation Facility (ESRF), Grenoble, France.
   index: 4
 - name: Astral Software Inc
   index: 5
 - name: NIST Center for Neutron Research
   index: 6
 - name: Institute of Geosciences, Meteorology Section, University of Bonn, Germany.
   index: 7
 - name: No affiliation.
   index: 8

date: 14 December 2025
bibliography: paper.bib

---

# Summary

`pyfive` is an open-source thread-safe pure Python package for reading data stored in HDF5. 
While it is not a complete implementation of all the specifications and capabilities of HDF5, it includes all the core functionality necessary to read gridded datasets, whether stored contiguously or with chunks (with or without standard compression options).
All data access is fully lazy, the data is only read from storage when the numpy data arrays are manipulated. Originally developed some years ago, the package has recently been upgraded to support lazy access, and to add missing features necessary for handling all the environmental data known to the authors. 

> It is now a realistic option for production data access in environmental science and more widely. 

Reviewer: ```suggestion
It is now a realistic option for production data access in environmental science and more broadly within other domains. 
```

The API is based on that of `h5py` ([https://github.com/h5py/h5py](https://github.com/h5py/h5py), a Python shimmy over the HDF5 C-library which itself is not thread-safe), with some API extensions to help optimise remote access. 
With these extensions, coupled with thread safety, many of the limitations precluding the efficient use of HDF5 (and netCDF4) on cloud storage have been removed.

# Statement of need

HDF5[^1][@FolEA11] is probably the most important data format in physical science.
It is particularly important in environmental science given the fact that netCDF4[^2][@Rew06] is HDF5 under the hood. 
From satellite missions, to climate models and radar systems, the default binary format has been HDF5 for decades. 

> While newer formats are starting to get mindshare, there are petabytes, if not exabytes, of existing HDF5, and there are still many good use-cases for creating new data in HDF5. 

Reviewer: ```suggestion
While newer data formats are starting to get mindshare, there are petabytes, if not exabytes, of existing HDF5, and there remain many viable reasons for creating new data in the HDF5 format today. 
```

> However, despite the history, there are few libraries for reading HDF5 file data that do not depend on the official HDF5 library maintained by the HDF Group, and in particular, apart from `pyfive`, in Python there are none that cover the needs of environmental science. 

Reviewer: ```suggestion
However, despite its historical importance, there are few libraries available for reading the HDF5 data format that do not depend on the official HDF5 library maintained by the HDF Group; Particularly for Python, apart from `pyfive`, there are no HDF5 libraries that cater to the needs of environmental science. 
```

Finally, the code complexity also meant that it is not suitable for developing bespoke code for data recovery in the case of partially corrupt data. 
From a long-term curation perspective both of these last two constraints are a concern. 

The issues of the dependency on a complex code maintained by one private company in the context of maintaining data access (over decades, and potentially centuries), can only be mitigated by ensuring that the data format is well documented, that data writers use only the documented features, and that public code exists which can be relatively easily maintained. 
The HDF5group have provided good documentation for HDF5, but while there is a community of developers beyond the HDF5 group, recent events suggest that given most of those developers and their existing funding are US based, some spreading of risk would be desirable. 
To that end, a pure Python code covering the core HDF5 features of interest to the target scientific community, which is relatively small and maintained by an international constituency provides some assurance that the community can maintain HDF5 access for the foreseeable future. 

> A pure Python code also makes it easier to develop scripts which can work around data and metadata damage should they occur, and has the additional advantage of being able to be deployed in resource or operating-system constrained environments (such as mobile).

Reviewer: ```suggestion
A library written entirely in Python code also makes it easier to develop scripts which can work around data and metadata corruption should they occur, and has the additional advantage of being easily deployed in resource or operating-system constrained environments (such as on mobile).
```

[^1]: https://www.hdfgroup.org/solutions/hdf5/
[^2]: https://www.unidata.ucar.edu/software/netcdf

# Current Status of pyfive

The original implementation of `pyfive` (by JH), which included all the low-level functionality to deal with the internals of an HDF5 file was developed with POSIX access in mind. 

> The recent upgrades were developed with the use-cases of performant remote access to curated data as the primary motivation - including full support for lazy loading only parts of chunked datasets as they are needed.

Reviewer: ```suggestion
The recent upgrades were developed with the use-cases of performant remote access to curated data as the primary motivation - including full support for lazy loading only the relevant parts of chunked datasets, as they are needed.
```

Thread safety has become a concern given the wide use of Dask[^3] in Python based analysis workflows, and this, coupled with a lack of user knowledge about how to efficiently use HDF5, has led to a community perception that HDF5 is not fit for remote access (especially on cloud storage).
``pyfive`` addresses thread safety by bypassing the underlying HDF5 c-library and
addresses some of the issues with remote access by optimising access to
internal file metadata (in particular, the chunk indexes) and by supporting the determination of whether or not a given file is cloud optimised. 
 

> To improve internal metadata access, `pyfive` now supports several levels of "laziness" for instantating chunked datasets (variables). The default method preloads internal indices to make parallellism more efficient, but a completely lazy option without index loading is possible. Neither load data until it is requested.

Reviewer: ```suggestion
To improve internal metadata access, `pyfive` supports several levels of laziness for instantating chunked datasets (variables). The default method preloads internal indices to make parallellism more efficient, but a completely lazy option without index loading is also possible. Neither method loads data until it is requested.
```

It is now a realistic option for production data access in environmental science and more widely. 
The API is based on that of h5py (which is a python shimmy over the HDF5 c-library, and hence is not thread-safe), with some API extensions to help optimise remote access. 
With these extensions, coupled with thread safety, many of the limitations precluding the efficient use of HDF5 (and NetCDF4) on cloud storage have been removed.

# Statement of need

HDF5[^1][@FolEA11] is probably the most important data format in physical science, used across the piste. 
It is particularly important in environmental science, particularly given the fact that NetCDF4[^2][@Rew06] is HDF5 under the hood. 
From satellite missions, to climate models and radar systems, the default binary format has been HDF5 for decades. 
While newer formats are starting to get mindshare, there are petabytes, if not exabytes, of existing HDF5, and there are still many good use-cases for creating new data in HDF5. 

> However, despite the history, there are few libraries for reading HDF5 file data that do not depend on the official HDF5 library maintained by the HDFGroup, and in particular, apart from pyfive, in Python there are none that cover the needs of environmental science. 

Reviewer 2: ```suggestion
However, despite the history, there are few libraries for reading HDF5 file data that do not depend on the official HDF5 library maintained by the HDF Group, and in particular, apart from pyfive, in Python there are none that cover the needs of environmental science. 
```

While the HDF5 c-library is reliable and performant, and battle-tested over decades, there are some caveats to depending upon it: 
Firstly, it is not thread-safe,  secondly, the code is large and complex, and should anything happen to the financial stability of The HDF5group, it is not obvious the C-code could be maintained. 
Finally, the code complexity also meant that it is not suitable for developing bespoke code for data recovery in the case of partially corrupt data. 
From a long-term curation perspective both of these last two constraints are a concern. 

[^1]: https://www.hdfgroup.org/solutions/hdf5/
[^2]: https://www.unidata.ucar.edu/software/netcdf

> The original implementation of `pyfive` (by JH), which included all the low-level functionality to deal with the internals of an HDF5 file was developed with POSIX access in mind. 

Reviewer 2: JH -> expand?

With these tools, index extraction with `pyfive` can be comparable in performance to obtaining a kerchunk index, and completely opaque to the user.

With the use of pyfive, suitably repacked and rechunked HDF5 data can now be considered "cloud-optimised", insofar as with lazy loading, 
improved index handling, and thread-safety, there are no "format-induced" constraints on performance during remote access.
To aid in discovering whether or not a given HDF5 dataset is cloud-optimised, `pyfive` also now provides simple methods to expose information about file layout - both in API extensions, and in a new `p5dump` utility packaged with the `pyfive` library,
which provides (in the default view) functionality similar to ncdump, and when used with `p5dump -s`, information about storage characteristics.

The issues of the dependency on a complex code maintained by one private company in the context of maintaining data access (over decades, and potentially centuries), can only be mitigated by ensuring that the data format is well documented, that data writers use only the documented features, and that public code exists which can be relatively easily maintained. 

> The HDF5group have provided good documentation for the core features of HDF5 which include all those of interest to the weather and climate community who motivated this reboot of pyfive, and while there is a community of developers beyond the HDF5 group (including some at the publicly funded Unidata institution), recent events suggest that given most of those developers and their existing funding are US based, some spreading of risk would be desirable. 

Reviewer 3: ```suggestion
The HDF5group have provided good documentation for the core features of HDF5 which include all those of interest to the weather and climate community who motivated this reboot of `pyfive`, and while there is a community of developers beyond the HDF5 group (including some at the publicly funded Unidata institution), recent events suggest that given most of those developers and their existing funding are US based, some spreading of risk would be desirable. 
```

A pure python code also makes it easier to develop scripts which can work around data and metadata damage should they occur. 

## Examples

We now introduce three aspects of the new functionality that `pyfive` now provides: remote access, configurable lazy loading, and determining whether files are cloud optimised.

##  Remote Access

A notable feature of the recent `pyfive` upgrade is that it was carried out with thread-safety and remote access using fsspec (filesystem-spec.readthedocs.io) in mind.  We provide two examples of using `pyfive` to access remote data, one in S3, and one behind a modern http web server:

For accessing the data on S3 storage, we will have to set up an ``s3fs`` virtual file system, then pass it to Pyfive:

```python
import pyfive
import s3fs
# storage options for an anon S3 bucket
# there are also caching options for the s3 middleware, not shown here
storage_options = {
    'anon': True,
    'client_kwargs': {'endpoint_url': "https://s3server.ac.uk"}
}
fs = s3fs.S3FileSystem(**storage_options)
file_uri = "s3-bucket/myfile.nc"

> with fs.open(file_uri, 'rb') as s3_file:

Reviewer 2: PEP 8
```suggestion
    "anon": True,
    "client_kwargs": {"endpoint_url": "https://s3server.ac.uk"}
}
fs = s3fs.S3FileSystem(**storage_options)
file_uri = "s3-bucket/myfile.nc"
with fs.open(file_uri, "rb") as s3_file:
```

    nc = pyfive.File(s3_file)
    dataset = nc[var]
```

for an HTTPS data server, the usage is similar:

```python
import fsspec
import pyfive
# there are also caching options for the fsspec middleware, not shown here
fs = fsspec.filesystem('http')
with fs.open("https://site.com/myfile.nc", 'rb') as http_file:
    nc = pyfive.File(http_file)
    dataset = nc[var]
```

This is of course exactly the same pattern as remote access using `h5py`, and that is by design - to make
moving to `pyfive` easy for users!

## Lazy Loading

A key tenet of efficient remote access is that variable inspection is quick and involves the minimimum of network
traffic between storage and the client application.  
However when this is coupled with the common pattern of 
using Dask, some flexibility in what is loaded when is beneficial.

> By default when one inspects the contents of a file using `pyfive v1.0` nothing more is read from the file than

Reviewer 2: Not sure we need to mention the version, here. I think we can take is as read that we're describing the latest and greatest.

```suggestion
By default when one inspects the contents of a file using `pyfive` nothing more is read from the file than
```

the names of the variables ("datasets" in the language of HDF5): for example:
```python
with pyfive.File('myfile.h5','r') as f:
  variables_in_file = [v for v in f]
```
involves nothing more than getting a set of variable names. When one wishes to inspect these variables:
``` python
with pyfive.File('myfile.h5','r') as f:

>   temp = f['temp']

Reviewer 2: ```suggestion
with pyfive.File("myfile.h5","r") as f:
    temp = f["temp"]

```

this point the b-tree index is also loaded and the variable can now be accessed outside the context
manager. 
Data loading is now completely lazy and the variable instance (`temp`) has all the information
needed to extract data as needed. 
This is done so that in Dask applications, when one passes
each Dask computational chunk a portion of the variable, each such Dask chunk has already got the 
index, and when it does want to load data, it can be as efficient as possible.

However, there are situations where loading the b-tree at variable instantation is not wanted, and all
is wanted is to be able to view all the variable attributes. 
To support this option `pyfive` also offers the `get_lazy_view` file method, so one can do:
``` python
with pyfive.File('myfile.h5','r') as f:
  temp = f.get_lazy_view('temp')
print(temp.attrs())
```

It is still possible to access the data and the b-tree index is loaded when data access is first attempted.

This extra lazy view is new functionality. 
It would obviously have been possible to make the default not load the b-tree, but in the opinion of the

> current `pyfive` maintainers loading the b-tree at variable instantiation is likely more consistent

Reviewer 2: ```suggestion
It would obviously have been possible to make the default not load the internal file metadata, but in the opinion of the
current `pyfive` maintainers, loading the internal file metadata during variable instantiation is likely more consistent
```

In the following example we compare and contrast the unpacked and repacked version of a particularly pathological file, and in doing so howcase some of the `pyfive` API extensions which help us understand why it is pathological, and how to address those issues for repacking.

If we extract just a piece of the output of `p5dump -s` on this file (which has surface wind velocity at three hour intervals for one hundred years):
```
float64 time(time) ;
                time:standard_name = "time" ;
                time:_n_chunks = 292192 ;
                time:_chunk_shape = (1,) ;
                time:_btree_range = (31808, 19854095942) ;
                time:_first_chunk = 9094 ;

float32 uas(time, lat, lon) ;
                uas:_Storage = "Chunked" ;
                uas:_n_chunks = 292192 ;
                uas:_chunk_shape = (1, 143, 144) ;
                uas:_btree_range = (28672, 19854809382) ;
                uas:_first_chunk = 36520 ;
```

> we can immediately see that this will be a problematic file!  The b-tree index is clearly interleaved with the data (compare the first chunk address with last index addresses of the two variables), and with a chunk dimension of (1,), any effort

Reviewer 2: ```suggestion
we can immediately see that this will be a problematic file! The internal file metadata index is clearly interleaved with the data (compare the first chunk address with last index addresses of the two variables), and with a chunk dimension of (1,), any effort
```

> to use the time-dimension to locate data of interest will involve a ludicrous number of 1 number reads (all underying

Reviewer 3: ```suggestion
to use the time-dimension to locate data of interest will involve a ludicrous number of 1 number reads (all underlying
```
